tket/src/Gate/Gate.cpp

```diff
@@ -221,13 +225,79 @@ Op_ptr Gate::transpose() const {
+  static const std::regex nan_regex("\\bnan\\b");
+  for (const Expr& e : params) {
+    stringstream ss;
+    ss << e;
+    if (std::regex_search(ss.str(), nan_regex)) return true;
+  }
+  return false;
+}
+
+static double random_perturbation() {
+  static RNG rng;
+  int a = rng.get_size_t(20);
+  return (a - 10) * EPS;
+}
+
+// If `to_doubles` is true, when an expression contains no free symbols evaluate
+// it as a double. This is appropriate in the case where an expression has been
+// perturbed, when there is no point retaining exact values for symbolic
+// constants.
+static std::vector<Expr> subs_all_params(
+    const std::vector<Expr>& params, const SymEngine::map_basic_basic& sub_map,
+    bool to_doubles = false) {
+  std::vector<Expr> new_params;
+  for (const Expr& p : params) {
+    Expr psub = p.subs(sub_map);
+    if (to_doubles && expr_free_symbols(psub).empty()) {
+      new_params.push_back(SymEngine::eval_double(psub));
+    } else {
+      new_params.push_back(psub);
+    }
+  }
+  return new_params;
+}
+
 Op_ptr Gate::symbol_substitution(
     const SymEngine::map_basic_basic& sub_map) const {
-  std::vector<Expr> new_params;
-  for (const Expr& p : this->params_) {
-    new_params.push_back(p.subs(sub_map));
+  // Perform symbolic substitution, but catch the case where the returned
+  // expression is not a number, and in that case try to set a value by
+  // perturbing the inputs. This deals with cases where expressions contain
+  // terms that are undefined at specific values but where the singularity is
+  // removable at the op level. (Non-removable singularities, which ought
+  // strictly to fail substitution, will result in spectacularly wrong values,
+  // but are unlikely to arise in practice.)
+  //
+  // This is a partial workaround for issues arising from squashing symbolic
+  // rotations, where it is hard or impossible to handle special cases (e.g.
+  // where the general formula reduces to one involving atan2(0,0)).
+  //
+  // A proper solution to this problem may have to wait for TKET 2.
+  std::vector<Expr> new_params = subs_all_params(this->params_, sub_map);
+  if (!params_contain_nan(new_params)) {  // happy path
+    return get_op_ptr(this->type_, new_params, this->n_qubits_);
   }
-  return get_op_ptr(this->type_, new_params, this->n_qubits_);
+
+  // Try perturbing all values in the map. May need several attempts in case
+  // there are subexpressions on the boundary of validity, such as acos(1.0). If
+  // we fail after 1000 attempts, give up.
+  for (unsigned i = 0; i < 1000; i++) {
+    SymEngine::map_basic_basic new_sub_map;
+    for (const auto& pair : sub_map) {
+      new_sub_map[pair.first] = Expr(pair.second) + random_perturbation();
+    }
+    std::vector<Expr> new_params_1 =
+        subs_all_params(this->params_, new_sub_map, true);
+    if (!params_contain_nan(new_params_1)) {
+      return get_op_ptr(this->type_, new_params_1, this->n_qubits_);
```

Comment:
I presume that forcing the reduction to a double here is how we enforce any `atan2` to decide whether to reduce to a `nan` or not. Since this will only happen when the full expression contains no symbols, I don't think this will always catch it when multiple symbols are instantiated one at a time. For example, if the expression is of the form `alpha + ... atan2(... beta ..., ...) ...` and `beta` is instantiated first, it could be set to a value that would yield a `nan` from the `atan2`. However, we would only detect this at the later point when `alpha` is instantiated, at which point the permutations would only affect `alpha` and so wouldn't resolve the `nan`. Perhaps we just let this case hit the exception? The other alternative I can think of is replacing the `eval_double` call with a more expensive call to the generic simplification procedure so we spot the `nan` at the point where a relevant symbol is instantiated. Which do you think would work better?

Reply:
I think this can be an exceptional case. It seems to be quite hard to trigger! (I tried.)
Another thing that can happen is infinities, which are different, and represented in symengine expressions by `zoo` etc. For example these arise when dividing by zero. We could expand the regex to catch these as well, but I think they will be even more rare in practice.

Reply:
Wouldn't the following example trigger it?
```
circ = Circuit(1)
circ.Rz(gamma, 0)
circ.Rx(alpha, 0)
circ.Ry(beta, 0)
circ.Rz(gamma, 0)
SquashTK1().apply(circ)
circ.symbol_substitution({alpha : 0., beta : 0.})
circ.symbol_substitution({gamma : 0.})
```
In between the two substitutions, the command is `TK1(1/2 - (3/4 - (1/2 + gamma) + nan*pi**(-1)), 0, -1/2 - (1/4 - (1/2 + gamma) + nan*pi**(-1))) q[0];` where both expressions containing `nan` also contain `gamma`.

But yeah, I'm happy for this to fall through to the exception. We shouldn't be generating the infinite cases within our compilation so it makes sense for that kind of thing to also hit the exception as well, since it should be traceable back to a user input error

Reply:
Thanks for the example, yes it should do!